tests/python/pants_test/base/test_hash_utils.py

```diff
@@ -82,3 +83,66 @@ def check_bad_spec(spec):
     check_bad_spec('/')
     check_bad_spec('foo/1')
     check_bad_spec('1/foo')
+
+
+class JsonEncodingTest(unittest.TestCase):
+  def _perform_json_dump(self, o, digest=None):
+    return _json_encode_for_hash(o, encoder=CoercingEncoder)
```

Comment:
OK - still not buying it, but not a blocking objection. If you want to test the CoercingEncoder, test the CoercingEncoder directly using json.dumps prhaps here in the test instead of creating the otherwise un-needed `_json_encode_for_hash`.

Reply:
I believe the reason I left the function (I removed it several commits ago and added it back) was because I thought the specific set of arguments for `json.dumps()` were necessary to ensure the "stable" part of `stable_json_hash()`, but there's no e.g. `stable_json_dumps()` method, because the intermediate json representation isn't something we expose publicly (and I don't think that we need to). I want to have tested `object -> json` conversion, mostly, and the checksum testing then just verifies that `json -> hex digest` works correctly.

To remove the import of the private method, I'm thinking I can make a public object representing the keyword arguments we pass to `json.dumps()` in `hash_utils.py`, and use that in our `stable_json_hash()` method and in our tests here. But to me, it seems extremely reasonable to test the *stable* json conversion (a property we rely upon!) in isolation, while explicitly marking it as an implementation detail and not to be used by itself. To me, making an explicit entrypoint to stable json conversion is reasonable even though it results in the import of a private method.

Like I said above, I removed it earlier, and couldn't find a way around it -- I would like to know if:
1. testing the `object -> json` conversion makes sense here (this is why i was unable to figure out how to remove `_json_encode_for_hash()`).
2. how best to test that and ensure we have the same behavior (using a keyword arguments object as above, or am i overthinking this?)

Reply:
I think you're overthinking.

If you want to test intermediate json, use the encoder directly in the test with json.dumps.
If you want to test stable hashes, test the opaque public hashing function directly in the test.

Reply:
After reading your comment and agreeing/realizing that stability is definitely a function of the hash, not of the intermediate json, I can absolutely remove that method and test as you describe here.

Reply:
Done!